src/Lean/Meta/Tactic/FunInd.lean

```diff
@@ -0,0 +1,855 @@
+where
+  go (packedDomain domain : Expr) args : MetaM Expr :=  do
+    if domain.isAppOfArity ``PSigma 2 then
+      let #[a, b] := domain.getAppArgs | unreachable!
+      withLocalDecl `x .default a fun x => do
+        mkLambdaFVars #[x] (← go packedDomain (b.beta #[x]) (args.push x))
+    else
+      withLocalDecl `x .default domain fun x => do
+        let args := args.push x
+        let packedArg ← WF.mkUnaryArg packedDomain args
+        mkLambdaFVars #[x] (e.beta #[packedArg])
+
+/--
+Iterated `PSigma.casesOn`: Given `y : a ⊗' b ⊗' …` and a type `codomain`,
+and `alt : (x : a) → (y : b) → codomain`, uses `PSigma.casesOn` to invoke `alt` on `y`.
+
+This very is similar to `Lean.Predefinition.WF.mkPSigmaCasesOn`, but takes a lambda rather than
+free variables.
+-/
+partial def mkPSigmaNCasesOn (y : FVarId) (codomain : Expr) (alt : Expr) : MetaM Expr := do
+  let mvar ← mkFreshExprSyntheticOpaqueMVar codomain
+  let rec go (mvarId : MVarId) (y : FVarId) (ys : Array Expr) : MetaM Unit := mvarId.withContext do
+    if (← inferType (mkFVar y)).isAppOfArity ``PSigma 2 then
+      let #[s] ← mvarId.cases y | unreachable!
+      go s.mvarId s.fields[1]!.fvarId! (ys.push s.fields[0]!)
+    else
+      let ys := ys.push (mkFVar y)
+      mvarId.assign (alt.beta ys)
+  go mvar.mvarId! y #[]
+  instantiateMVars mvar
+
+/--
+Given expression `e` with type `(x : A) → (y : B[x]) → … → (z : D[x,y]) → R`
+return an expression of type `(x : A ⊗' B ⊗' … ⊗' D) → R`.
+-/
+partial def curryPSigma (e : Expr) : MetaM Expr := do
+  let (d, codomain) ← forallTelescope (← inferType e) fun xs codomain => do
+    if xs.any (codomain.containsFVar ·.fvarId!) then
+      throwError "curryPSum: codomain depends on domain variables"
+    let mut d ← inferType xs.back
+    for x in xs.pop.reverse do
+      d ← mkLambdaFVars #[x] d
+      d ← mkAppOptM ``PSigma #[some (← inferType x), some d]
+    return (d, codomain)
+  withLocalDecl `x .default d fun x => do
+    let value ← mkPSigmaNCasesOn x.fvarId! codomain e
+    mkLambdaFVars #[x] value
+
+/--
+Given type `(a ⊗' b ⊕' c ⊗' d) → e`, brings `a → b → e` and `c → d → e`
+into scope and passes them to the contiuation.
+The `name` is used to form the variable names; uses `name1`, `name2`, … if there are multiple.
+-/
+def withCurriedDecl {α} (name : String) (type : Expr) (k : Array Expr → MetaM α) : MetaM α := do
+  let some (d,c) := type.arrow? | throwError "withCurriedDecl: Expected arrow"
+  let motiveTypes ← (unpackPSum d).mapM (uncurryPSumArrow · c)
+  if let [t] := motiveTypes then
+    -- If a singleton, do not number the names.
+    withLocalDecl name .default t fun x => do k #[x]
+  else
+    go motiveTypes #[]
+where
+  go : List Expr → Array Expr → MetaM α
+  | [], acc => k acc
+  | t::ts, acc => do
+    let name := s!"{name}{acc.size+1}"
+    withLocalDecl name .default t fun x => do
+      go ts (acc.push x)
+
+
+/--
+Given expression `e` of type `(x : a ⊗' b ⊕' c ⊗' d) → e[x]`, returns expression of type
```

Comment:
This isn't `e`'s type, is it?

I generally don't understand what this docstring is indicating.